tests/test_auth.py

```diff
@@ -162,3 +162,42 @@ def test_get_password_runtime_error_suppressed(
     assert len(recwarn) == 1
     warning = recwarn.pop(UserWarning)
     assert "fail!" in str(warning)
+
+
+def test_get_username_raises_attribute_error(entered_username, monkeypatch, config):
+    """Test when Keyring's get_credential raises an AttributeError"""
+
+    class FailKeyring:
+        @staticmethod
+        def get_credential(system, username):
+            raise AttributeError("fail!")
+
+    monkeypatch.setattr(auth, "keyring", FailKeyring())
+    assert auth.Resolver(config, cred()).username == "entered user"
+
+
+def test_get_username_raises_exception(entered_username, monkeypatch, recwarn, config):
+    """Test when Keyring's get_credential raises an Exception"""
+
+    class FailKeyring:
+        @staticmethod
+        def get_credential(system, username):
+            raise Exception("fail!")
+
+    monkeypatch.setattr(auth, "keyring", FailKeyring())
+    assert auth.Resolver(config, cred()).username == "entered user"
+    assert len(recwarn) == 1
+    warning = recwarn.pop(UserWarning)
+    assert "fail!" in str(warning)
```

Comment:
I just ran `tox -e py` locally on the latest master, and it shows that `twine/auth.py` has a test coverage of 98%, with the branch from `59->66` not fully covered, which I covered by adding these tests.

```
$ git rev-parse --abbrev-ref HEAD
master

$ twine --version
twine version 3.1.2.dev96+g895ac8a (pkginfo: 1.5.0.1, requests: 2.23.0, setuptools: 41.2.0, requests-toolbelt: 0.9.1, tqdm: 4.45.0)

==================================================================================================== 135 passed, 47 warnings in 42.72s =====================================================================================================
py run-test: commands[1] | coverage report -m
Name                         Stmts   Miss Branch BrPart  Cover   Missing
------------------------------------------------------------------------
twine/__init__.py               14      1      2      1    88%   29->32, 32
twine/_installed.py             42      4     18      7    82%   20->57, 22->23, 23, 27->57, 46->57, 47->52, 49->53, 52, 53->46, 57-60
twine/auth.py                   57      0      4      1    98%   59->66
twine/cli.py                    29      0      4      0   100%
twine/commands/__init__.py      21      0     11      0   100%
twine/commands/check.py         73      0     22      1    99%   98->105
twine/commands/register.py      25      0      4      0   100%
twine/commands/upload.py        58      2     22      2    95%   83->84, 84, 85->86, 86
twine/exceptions.py             27      0      0      0   100%
twine/package.py               130      8     35      8    90%   82->87, 87, 100->101, 101-102, 105->106, 106, 164->165, 165, 168, 240->exit, 244->246, 246, 249->exit, 253->255, 255
twine/repository.py            124      1     38      0    99%   257
twine/settings.py               71      0      6      0   100%
twine/utils.py                 107      0     38      1    99%   191->196
twine/wheel.py                  50      0     18      0   100%
twine/wininst.py                38     26     19      0    21%   15-17, 21-25, 28-59
------------------------------------------------------------------------
TOTAL                          866     42    241     21    93%
py run-test: commands[2] | coverage html
_________________________________________________________________________________________________________________ summary __________________________________________________________________________________________________________________
  py: commands succeeded
  congratulations :)
```

Reply:
Ah. The `59->66` indicates that the jump from line 59 to 66 wasn't covered: https://coverage.readthedocs.io/en/coverage-5.1/branch.html. If the whole block wasn't covered, it would be something like `59-66`.

I verified this by skipping the tests that are indicated in blue in the screenshot above. That resulted in:

```
twine/auth.py                   57      5      4      1    87%   59->66, 61-66
```

<img width="730" alt="Screen Shot 2020-05-17 at 10 53 01 AM" src="https://user-images.githubusercontent.com/1326704/82152042-c9368d80-982c-11ea-9f8f-02f7e8c22522.png">

Note here that I'm hovering over the yellow "partially covered" line, which explains what's missing.

Compare that to the screenshot above, where lines 61 through 66 are green, indicating they were covered.

So, my request to remove these tests still stands. 😉

Reply:
Yes, the jump from 59->66 wasn't covered, and the added tests covered that branch and brought the test coverage to 100%.

I was under the assumption that tests if we are using `coverage` to track branch coverage, there should be tests added to cover those branches. In-fact, in few of the previous unit test PRs I added, there were tests covering those branches, and the tests I added cover  the branches here by explicitly raising those exceptions in `FailKeyring.get_credentials`

https://github.com/pypa/twine/blob/895ac8af383b99d9b1673f2d7edcea92e4b5a429/twine/auth.py#L59-L66

I am happy to apply any improvements to these tests if suggested, like parametrisation etc as well. Or I can go ahead and remove the tests, but then the coverage for `twine.auth` won't hit a 100%, which I thought is the aim to achieve when writing unit tests, if you can hit that 100% mark.